src/services/cache-service.ts

```diff
@@ -2,63 +2,110 @@
  * Caches the response of any promise value of a loading function and prevents duplicate calls
  * @author Mike Reiche <mike.reiche@t-systems.com>
  */
+import {TimeComponents, toMilliseconds} from "../utils/time";
+
 export class CacheService {
-    private _cacheContainer = {};
-    private _defaultCacheTtlSeconds = 10;
+    private _cacheContainer:{[key: string]: [number, Promise<any>]} = {};
+    private _defaultCacheTtl:TimeComponents = {seconds: 10};
 
+    /**
+     *
+     * @param key The key for cached value
+     * @param loadingFunction The loading function returning a promise of the value.
+     * @param ttl The cache timeout for the response.
+     */
     getForKeyWithLoadingFunction <T> (
         key:string,
         loadingFunction:() => Promise<T>,
-        cacheTtlSeconds = this._defaultCacheTtlSeconds
+        ttl:number|TimeComponents = this._defaultCacheTtl
     ):Promise<T> {
-        let cacheEntry = this._cacheContainer[key];
-        const now = (Date.now()/1000);
-        if (cacheEntry !== undefined) {
-            if (cacheEntry[0] >= now - cacheTtlSeconds) {
-                return cacheEntry[1];
-            }
+        const cacheEntry = this._cacheContainer[key];
+        const now = Date.now();
+
+        if (cacheEntry !== undefined && cacheEntry[0] >= now) {
+            return cacheEntry[1];
         }
 
+        ttl = CacheService.wrapTtl(ttl);
+        const cacheTtlMs = toMilliseconds(ttl);
+        const validUntil = now + cacheTtlMs;
         this._cacheContainer[key] = [
-            now,
-            loadingFunction().then(object=>{
-                return this._cacheContainer[key][1] = Promise.resolve(object);
```

Comment:
Alas, I fear my daftness knows no bounds today. There _is_ an issue, and even _this_ issue, it simply manifests differently from how I thought it would.

Here is a test case:
```typescript
test("Cached value cannot be dirty-read", async () => {
  const cacheService = new CacheService()
  const key = "someKey"
  // Each must be greater than the previous one. Script may run too slowly for the math to work out for two-digit values.
  const ttl = {ms: 100}
  const cleanupTime = {ms: 150}
  const loadingTime = {ms: 200}
  // afterLoadingTimeButBeforeTtlEnds must be less than ttl.
  const afterLoadingTimeButBeforeTtlEnds = {ms: 1 + loadingTime.ms - cleanupTime.ms}
  expect(afterLoadingTimeButBeforeTtlEnds.ms).toBeLessThan(ttl.ms) // Sanity Check

  let counter = 0
  const fastLoadingFunction = () => Promise.resolve(counter++)
  const slowLoadingFunction = async () => {
    const result = counter++ // Copy-by-Value before another call can change counter
    await sleep(toMilliseconds(loadingTime))
    return Promise.resolve(result)
  }

  const actualResults = new Array<number>()
  cacheService.getForKeyWithLoadingFunction(key, slowLoadingFunction, ttl, 0)
      .then(result => actualResults[0] = result)

  await sleep(toMilliseconds(cleanupTime))
  cacheService.invalidateOutdated() // entry of slowLoadingFunction removed

  // This will find no entry and thence call fastLoadingFunction.
  cacheService.getForKeyWithLoadingFunction(key, fastLoadingFunction, ttl, 1)
      .then(result => actualResults[1] = result)

  // This will find the valid entry of fastLoadingFunction and not call fastLoadingFunction again.
  cacheService.getForKeyWithLoadingFunction(key, fastLoadingFunction, ttl, 2)
      .then(result => actualResults[2] = result)

  // After this wait, the entry from fastLoadingFunction is still valid - but has just been replaced by the zombie key from slowLoadingFunction.
  await sleep(toMilliseconds(afterLoadingTimeButBeforeTtlEnds))
  const keys = cacheService.outdatedKeys

  // This will find the zombie entry written by resolving slowLoadingFunction instead of the valid entry written by fastLoadingFunction.
  const lastCall = cacheService.getForKeyWithLoadingFunction(key, fastLoadingFunction, ttl, 3)
      .then(result => actualResults[3] = result)
  await lastCall // extra line so IDEA breakpoint on previous line works normally

  let previousResult = -1
  for (const result of actualResults) { // Sanity Check: no past value read in the future
    expect(result).toBeGreaterThanOrEqual(previousResult)
    previousResult = result
  }
  expect(actualResults[2]).toBe(actualResults[1]) // Call 2 did not happen because entry 1 is still valid.
  expect(actualResults[3]).toBe(actualResults[1]) // Call 3 should not have happened because entry 1 is still valid.
}, 1000000) // exaggeratedly high timeout so we can use breakpoints without breaking a sweat
```
Oh well, I merged it, I will prepare a fix. At least I noticed before release.

Opened #49 and assigned to me. Fix (and test case) incoming.